pages/blog/graphql-error-handling-with-fp.mdx

```diff
@@ -0,0 +1,640 @@
+import withArticle from '../../ui/blog/article'
+
+export const meta = {
+  title: 'GraphQL error handling to the max with Typescript, codegen and fp-ts',
+  author: 'Ghislain Thau',
+  tags: ['graphql', 'typescript', 'codegen', 'fp-ts'],
+  date: '2022-02-17',
+  description:
+    'Handle unsafe APIs in a typesafe way with Typescript and Functional Programming',
+}
+
+export default withArticle({ ...meta })
+
+Contrary to REST, GraphQL does not use Status Code to differentiate a successful result from an exception. In REST, one can send a response with a appropriate status code to inform the caller about the kind of exception encountered, e.g.:
+- 400: bad request (e.g. invalid inputs)
+- 401 or 403: unauthorized (need authentication) or forbidden (insufficient permission)
+- 404: resource not found
+- 500: internal server error
+- and more.
+
+
+# GraphQL default error handling
+
+
+In GraphQL, all requests must go to a single endpoint:
+
```

Comment:
You could still actually do that no?
Send 400 and other status codes, based on queries.
I'm not saying it's a good practice but it is possible and that also helps when using old school tools for monitoring that don't understand GraphQL

Reply:
Does it make sense to mention it in the context of this article?

Do you want me to reword it to make it clearer that it's customary but not mandatory to send 200 status codes for all queries returns?

Reply:
yes, I think it's important because many people are using this argument to bash on GraphQL compared to REST, saying GraphQL can't have normal status codes that work with popular monitoring tools.
But that's false, because you can treat GraphQL like REST and have all of that, but, you can also make it much better, with the approach you are suggesting.
So it's not a downside of GraphQL, it's something that GraphQL can do but can also do much better :)

Reply:
Looking at [graphql-helix implementation](https://github.com/contra/graphql-helix/blob/a796f9d2afdd8c2d91b745edbc53c5d49b716ba6/packages/core/lib/process-request.ts#L64), it seems that this must be handled out of the query resolvers code, at the server level, right?  One must throw an error from the query resolver (or not handle a bubbled up error), then at the server level catch and type check the possible errors to assign the "correct" http status code.

Another question: how is this managed client-side, e.g. with Apollo client? All non-200 response returns will be considered as errors? And in observable pipelines would need to be handled with an RXJS [catchError](https://rxjs.dev/api/operators/catchError) operator?

Reply:
Yes that's true, but also maybe in other implementation it is handled differently.
I wouldn't go too deep into it because there are many different ways to implement it, I would just mention it is possible.
About the client, also depends on the client you are using, so also here I would stay high level and general

Reply:
Ok, I'll phrase to mention that it's possible. And link to example of how graphql-helix does it?

Reply:
yes that's ok, we can always update it later with Yoga :)